File: src/main/python/searchA2DMatrix2/SearchA2DMatrix2.py
class Solution:
    def searchMatrix(self, matrix: list[list[int]], target: int) -> bool:
        for row in matrix:
            if row[0] > target:
                return False
            idx = 0
            while idx < len(row):
                if target == row[idx]:
                    return True
                if target < row[idx]:
                    break
                idx += 1
        return False

    def test(self):
        tests = [
            [
                [
                    [1, 4, 7, 11, 15],
                    [2, 5, 8, 12, 19],
                    [3, 6, 9, 16, 22],
                    [10, 13, 14, 17, 24],
                    [18, 21, 23, 26, 30],
                ],
                5,
                True,
            ],
            [
                [
                    [1, 4, 7, 11, 15],
                    [2, 5, 8, 12, 19],
                    [3, 6, 9, 16, 22],
                    [10, 13, 14, 17, 24],
                    [18, 21, 23, 26, 30],
                ],
                20,
                False,
            ],
        ]
        for test in tests:
            matrix, target, expect_result = test
            result = self.searchMatrix(matrix, target)
            print(result == expect_result)


if __name__ == "__main__":
    Solution().test()

File: src/main/kotlin/me/bossm0n5t3r/leetcode/searchA2DMatrix2/SearchA2DMatrix2.kt
package me.bossm0n5t3r.leetcode.searchA2DMatrix2

import java.util.LinkedList

class SearchA2DMatrix2 {
    class Solution {
        fun searchMatrix(
            matrix: Array<IntArray>,
            target: Int,
        ): Boolean {
            val queue = LinkedList<Pair<Int, Int>>()
            val m = matrix.size
            val n = matrix.first().size
            val visited = Array(m) { BooleanArray(n) { false } }
            queue.add(Pair(0, 0))
            visited[0][0] = true
            while (queue.isNotEmpty()) {
                val (r, c) = queue.poll()
                if (matrix[r][c] == target) return true
                if (r + 1 < m && !visited[r + 1][c] && matrix[r + 1][c] <= target) {
                    visited[r + 1][c] = true
                    queue.add(Pair(r + 1, c))
                }
                if (c + 1 < n && !visited[r][c + 1] && matrix[r][c + 1] <= target) {
                    visited[r][c + 1] = true
                    queue.add(Pair(r, c + 1))
                }
            }
            return false
        }

        fun searchMatrixEfficient(
            matrix: Array<IntArray>,
            target: Int,
        ): Boolean {
            if (matrix.isEmpty() || matrix.first().isEmpty()) return false
            var row = 0
            var col = matrix.first().size - 1
            while (col >= 0 && row < matrix.size) {
                val cur = matrix[row][col]
                if (cur == target) {
                    return true
                } else if (cur < target) {
                    row++
                } else {
                    col--
                }
            }
            return false
        }
    }
}
